uportal-war/src/main/java/org/jasig/portal/groups/pags/testers/AdHocGroupTester.java

```diff
@@ -0,0 +1,172 @@
+/**
+ * Licensed to Apereo under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work
+ * for additional information regarding copyright ownership.
+ * Apereo licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License.  You may obtain a
+ * copy of the License at the following location:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.jasig.portal.groups.pags.testers;
+
+import org.apache.commons.lang.Validate;
+import org.jasig.portal.EntityIdentifier;
+import org.jasig.portal.groups.IEntityGroup;
+import org.jasig.portal.groups.IGroupMember;
+import org.jasig.portal.groups.pags.IPersonTester;
+import org.jasig.portal.groups.pags.dao.IPersonAttributesGroupTestDefinition;
+import org.jasig.portal.portlets.groupselector.EntityEnum;
+import org.jasig.portal.security.IPerson;
+import org.jasig.portal.services.GroupService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+/**
+ * Immutable PAGS Tester for inclusive/exclusive membership in sets of groups. To avoid infinite recursion,
+ * calls to {@code test()} are tracked by parameters plus thread ID.
+ *
+ * @author  Benito J. Gonzalez <bgonzalez@unicon.net>
+ * @see     org.jasig.portal.groups.pags.IPersonTester
+ * @see     org.jasig.portal.groups.pags.dao.EntityPersonAttributesGroupStore
+ * @see     org.jasig.portal.groups.pags.dao.IPersonAttributesGroupTestGroupDefinition
+ * @since   4.3
+ */
+public final class AdHocGroupTester implements IPersonTester {
+
+    private static final Logger logger = LoggerFactory.getLogger(AdHocGroupTester.class);
+    private static final Set<String> currentPersons = new ConcurrentSkipListSet<>();
```

Comment:
The method that creates all these instances is initializeTester in EntityPersonAttributesGroupStore. During my tests, tester classes are continuously being created. Using EhCache here would possibly speed up performance given all the object creation. Testers appear to be immutable, so good candidates for caching. Further, we should consider caching at the group level, but we still need to account for using the same AdHocGroupTester object when the parameters are exactly the same.